modules/navier_stokes/src/fv/bcs/NSFVMomentumAdvectionFunctionBC.C

```diff
@@ -0,0 +1,149 @@
+    NSFVAdvectionBase(params),
+    _pressure_exact_solution(getFunction("pressure_exact_solution")),
+    _rho(getADMaterialProperty<Real>("rho")),
+    _mu(getADMaterialProperty<Real>("mu"))
+{
+}
+
+void
+NSFVMomentumAdvectionFunctionBC::interpolate(Moose::FV::InterpMethod m,
+                                             ADRealVectorValue & v_face,
+                                             const ADRealVectorValue & elem_v,
+                                             const RealVectorValue & ghost_v)
+{
+  Moose::FV::interpolate(
+      Moose::FV::InterpMethod::Average, v_face, elem_v, ghost_v, *_face_info, true);
+
+  if (m != Moose::FV::InterpMethod::RhieChow)
+    return;
+
+  auto tup = Moose::FV::determineElemOneAndTwo(*_face_info, *_p_var);
+  const Elem * const elem = std::get<0>(tup);
+  const bool elem_is_elem = std::get<2>(tup);
+  mooseAssert(elem_is_elem ? elem == &_face_info->elem() : elem == _face_info->neighborPtr(),
+              "elem_is_elem is incorrect");
+
+  const Point & elem_centroid =
+      elem_is_elem ? _face_info->elemCentroid() : _face_info->neighborCentroid();
+  const Real elem_volume = elem_is_elem ? _face_info->elemVolume() : _face_info->neighborVolume();
+
+  // Get pressure gradient for the elem
+  const VectorValue<ADReal> & grad_p_elem = _p_var->adGradSln(elem);
+
+  // Get pressure gradient for the ghost
+  RealVectorValue grad_p_ghost =
+      _pressure_exact_solution.gradient(_t, 2. * _face_info->faceCentroid() - elem_centroid);
+
+  // Uncorrected face pressure gradient
+  auto unc_grad_p = (grad_p_elem + grad_p_ghost) / 2.;
+
+  // Now perform correction
+
+  // Get pressure value for the elem
+  ADReal p_elem = _p_var->getElemValue(elem);
+
+  // Get pressure value for the ghost
+  Real p_ghost =
+      _pressure_exact_solution.value(_t, 2. * _face_info->faceCentroid() - elem_centroid);
+
+  auto d_cf = 2. * (_face_info->faceCentroid() - elem_centroid);
+  auto del_pressure = p_ghost - p_elem;
+
+  // Direction is important. We always define things to point out from the elem
+  if (!elem_is_elem)
+  {
+    d_cf *= -1;
+    del_pressure *= -1;
+  }
+  const auto d_cf_norm = d_cf.norm();
+  const auto e_cf = d_cf / d_cf_norm;
+
+  // Corrected face pressure gradient
+  const auto grad_p = unc_grad_p + (del_pressure / d_cf_norm - unc_grad_p * e_cf) * e_cf;
+
+  // Now we need to perform the computations of D
+  // I don't know how I would want to do computation of the a coefficient on a ghost cell. I would
+  // have to essentially create an entire fictional element with defined geometric locations of
+  // the faces in order to compute inward advective flux and diffusive flux. For now I'm going to
+  // try not doing that and just use the a coeff of the elem
+  const ADReal & face_a = rcCoeff(*elem, _mu[_qp], _rho[_qp]);
+  const Real face_volume = elem_volume;
+
+  const ADReal face_D = face_volume / face_a;
+
+  // perform the pressure correction
+  v_face -= face_D * (grad_p - unc_grad_p);
```

Comment:
this is very different from 15.111? It looks like you are using RC for in-domain elements, but adapted to the boundary

v_face should be elem_v - face_D * (grad_P(boundary) - grad_P(centroid) )
15.111 is much simpler to obtain too, no need to compute the ghost gradient or the effective FD gradient. 
The call to average interpolation at the beginning is not necessary too